use zap for logging

grounded042 opened the pull request:
This PR adds structured logging using zap for #55. I’ve gone around and cleaned up any usages of logrus or log to use zap. In my experience people have varying opinions on global loggers - I stuck with the philosophy that zap espouses which is to avoid global loggers when possible.

- For the CLI I didn’t switch anything to zap, though I did switch some places from using `log` to use the `github.com/fission/fission/fission/log` package.
- Some places that logged and returned an error have been switched to just return an error - in any place I no longer log in addition to returning an error I checked to make sure that any callers do log the returned error. I mostly did this in functions where I didn’t want to pass in a logger object.
- In any place I could clean up error wrapping or Sprintf-ing I did
- I found a couple of places where errors were swallowed and I just added a log message for those
- when testing it out in minikube I ran into an error with the fluentd image so I changed up the helm chart to have a separate value for the repository for the image
- I changed up the logging middleware to also use zap so you get structure all the time
- ~I updated the tests, but there were some inconsistencies between some tests and the implementations along with some client go errors from k8s~ resolved

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/fission/fission/1112)
<!-- Reviewable:end -->

soamvasani commented on canaryconfigmgr/canaryConfigMgr.go:
```diff
@@ -115,17 +122,21 @@ func (canaryCfgMgr *canaryConfigMgr) initCanaryConfigController() (k8sCache.Stor
 
 func (canaryCfgMgr *canaryConfigMgr) addCanaryConfig(canaryConfig *crd.CanaryConfig) {
-	log.Printf("addCanaryConfig called for %s", canaryConfig.Metadata.Name)
+	canaryCfgMgr.logger.Info("addCanaryConfig called", zap.String("called_for", canaryConfig.Metadata.Name))
```
shouldn't it be something like "canary_config" or even just "name" instead of "called_for"?

grounded042 replied:
yup, I'll fix that

soamvasani commented on canaryconfigmgr/canaryConfigMgr.go:
```diff
@@ -266,23 +344,21 @@ func (canaryCfgMgr *canaryConfigMgr) updateHttpTriggerWithRetries(triggerName, t
 		if err != nil {
-			log.Printf("Error getting http trigger object : %v", err)
-			return err
+			return errors.Wrap(err, "error getting http trigger object")
```
log removed, is this intentional?

grounded042 replied:
not intentional - I'll add back

soamvasani commented on canaryconfigmgr/canaryConfigMgr.go:
```diff
@@ -266,23 +344,21 @@ func (canaryCfgMgr *canaryConfigMgr) updateHttpTriggerWithRetries(triggerName, t
 		switch {
 		case err == nil:
-			log.Printf("Updated Http trigger : %s.%s", triggerName, triggerNamespace)
+			canaryCfgMgr.logger.Info("updated http trigger", zap.String("trigger", fmt.Sprintf("%s.%s", triggerName, triggerNamespace)))
```
hmm, aren't we supposed to have two zap.Strings, one for each of name and namespace?

grounded042 replied:
correct, I'll fix that up

soamvasani commented on canaryconfigmgr/canaryConfigMgr.go:
```diff
@@ -266,23 +344,21 @@ func (canaryCfgMgr *canaryConfigMgr) updateHttpTriggerWithRetries(triggerName, t
 			return nil
 		case k8serrors.IsConflict(err):
-			log.Printf("Conflict in updating http trigger : %s.%s, retrying", triggerName, triggerNamespace)
+			canaryCfgMgr.logger.Info("conflict in updating http trigger, retrying", zap.Error(err), zap.String("trigger", fmt.Sprintf("%s.%s", triggerName, triggerNamespace)))
```
same as above, replace sprintf with two zap.Strings

grounded042 replied:
:+1:

soamvasani commented on canaryconfigmgr/canaryConfigMgr.go:
```diff
@@ -266,23 +344,21 @@ func (canaryCfgMgr *canaryConfigMgr) updateHttpTriggerWithRetries(triggerName, t
 		default:
-			log.Printf("Error updating trigger : %s.%s = %v", triggerName, triggerNamespace, err)
-			return err
+			return errors.Wrapf(err, "error updating trigger: %s.%s", triggerName, triggerNamespace)
```
log removed, is it intentional?

grounded042 replied:
nope, I'll add that back in - the thinking was this function was returning an error that would be logged at a higher level, but that's not the case.

soamvasani commented on canaryconfigmgr/canaryConfigMgr.go:
```diff
@@ -293,24 +369,31 @@ func (canaryCfgMgr *canaryConfigMgr) updateCanaryConfigStatusWithRetries(cfgName
 		if err != nil {
-			log.Printf("Error getting http Canary Config object : %v", err)
-			return err
+			return errors.Wrap(err, "error getting http canary config object")
```
log removed?

grounded042 replied:
same as above - I'll add it back

soamvasani commented on canaryconfigmgr/canaryConfigMgr.go:
```diff
@@ -293,24 +369,31 @@ func (canaryCfgMgr *canaryConfigMgr) updateCanaryConfigStatusWithRetries(cfgName
 		default:
-			log.Printf("Error updating Canary Config : %s.%s = %v", cfgName, cfgNamespace, err)
-			return err
+			return errors.Wrapf(err, "error updating canary config: %s.%s", cfgName, cfgNamespace)
```
log removed?

grounded042 replied:
same as above - I'll add back

grounded042 commented on builder/builder.go:
```diff
@@ -32,9 +30,11 @@ import (
 	"strings"
 	"time"
 
-	"github.com/dchest/uniuri"
```
will do

grounded042 commented on buildermgr/buildermgr.go:
```diff
@@ -17,29 +17,29 @@ limitations under the License.
 	err = fissionClient.WaitForCRDs()
 	if err != nil {
-		log.Fatalf("Error waiting for CRDs: %v", err)
+		return errors.Wrap(err, "error waiting for CRDs")
```
This was changed to align with the code in https://github.com/fission/fission/blob/2f42b795263177fe83c3417439e1ea7e7a6608ac/fission-bundle/main.go which on an error being returned `Fatal` is called. If you'd rather do a fatal log here, let me know and I can switch it up.

grounded042 commented on buildermgr/pkgwatcher.go:
```diff
@@ -86,20 +89,26 @@ func (pkgw *packageWatcher) build(buildCache *cache.Cache, srcpkg *crd.Package)
-		updatePackage(pkgw.fissionClient, srcpkg, fission.BuildStatusFailed, e, nil)
+		e := "error setting package pending state"
+		pkgw.logger.Error(e, zap.Error(err))
+		// I don't understand what the following line does - it seems like it's
```
I will remove it then.

grounded042 commented:
Thanks for the review @soamvasani! I've addressed the comments from above and I also went around and fixed up all the imports to reflect the three paragraph setup you mentioned.

grounded042 commented:
There seems to be an error in Travis that I am not able to reproduce locally.

grounded042 commented:
> There seems to be an error in Travis that I am not able to reproduce locally.

Fixed by rebasing with new changes from master

grounded042 commented:
Good catch @soamvasani. Let me change those checks up.

grounded042 commented:
Changed up and good to go @soamvasani!

vishal-biyani commented:
@grounded042 Can you please rebase and push once more.

grounded042 commented:
@vishal-biyani done. I also tried to remove the logrus package from glide using `glide remove github.com/sirupsen/logrus` but it did some really strange stuff to the lock file that then seemed to break the build. So if we want that removed maybe someone more versed in glide can do that. Otherwise, this is good to go.

vishal-biyani commented:
Thanks, @grounded042 I will cherry-pick your changes and run a full integration test in a separate PR. Once that is green - I will merge this one.  If at all needed to rebase once more, I will request before finally merging. Again thanks a lot for your contribution - excited to have this big change done!

grounded042 commented:
@vishal-biyani I cherry-picked it and pushed it up. Do you need me to rebase from master again? It looks like there are no merge conflicts.

vishal-biyani commented:
@grounded042 Yes you will have to rebase once please

grounded042 commented:
@vishal-biyani done

vishal-biyani commented:
The integration tests passed in https://github.com/fission/fission/pull/1118
